### DISCUSSION TITLE
Variable initialization in "if" statement

### DISCUSSION THREAD

**Discussion by foxesknow** [dotnet/csharplang#4332] [General]
Currently I can write code like this to see if a string exists within another string

```
static void Test(string s)
{
    if (s.IndexOf("a") is var index && index >= 0)
    {
       Console.WriteLine("bingo");
    }
}
```

However, I was surprised to see that if I change the code to:

```
static void Test(string s)
{
    if (s.IndexOf("a") is var index)
    {
       Console.WriteLine("bingo");
    }
}
```

It compiles and has the effect of meaning that the if expression is always true. i.e. it is treated as:
```

static void Test(string s)
{
    int num = s.IndexOf("a");
    if (true)
    {
        Console.WriteLine("bingo");
    }
}
```

This strikes be as a bug. The compiler should raise an error saying that the expression does not evaluate to a boolean.

**Comment by HaloFour** [comment]
> This strikes be as a bug. The compiler should raise an error saying that the expression does not evaluate to a boolean.

This is working as intended.  Pattern expressions do evaluate to a `bool` and the `var` pattern is always `true`.

**Reply by foxesknow** [reply]
Thanks.

This variation, without the var, also always evaluates to true:

```
static void Test(string s)
{
    if (s.IndexOf("a") is {} index)
    {
       Console.WriteLine("bingo");
    }
}
```
I'm guessing that this also works as intended, but it really doesn't look right!

**Reply by HaloFour** [reply]
The `{}` pattern is a little weird.  It's a degenerate form of the recursive property pattern, which normally looks like `TypeName { Property1: pattern1, ... }`.  However, the compiler can infer the type if you omit it, and you are allowed to have zero property patterns.  What that leaves you with is effectively a pattern that tests if the expression is of the inferred type, which is always true unless the expression is `null`.

TL;DR:  It's a weird pattern form of null-check.

**Comment by 333fred** [comment]
There is no bug here. The expression does evaluate to a bool, and the var pattern always matches the input value. The pattern you want is `s.IndexOf("a") is > 0`.
